src/components/app/routes/data/utils.js

```diff
@@ -60,46 +60,48 @@ export async function ensureEnterpriseAppData({
         const activatedOrAutoAppliedLicense = await activateOrAutoApplySubscriptionLicense({
           enterpriseCustomer,
           allLinkedEnterpriseCustomerUsers,
           subscriptionsData,
           requestUrl,
         });
         if (activatedOrAutoAppliedLicense) {
-          const { licensesByStatus } = subscriptionsData;
-          const updatedLicensesByStatus = { ...licensesByStatus };
-          Object.entries(licensesByStatus).forEach(([status, licenses]) => {
-            const licensesIncludesActivatedOrAutoAppliedLicense = licenses.some(
-              (license) => license.uuid === activatedOrAutoAppliedLicense.uuid,
-            );
-            const isCurrentStatusMatchingLicenseStatus = status === activatedOrAutoAppliedLicense.status;
-            if (licensesIncludesActivatedOrAutoAppliedLicense) {
-              updatedLicensesByStatus[status] = isCurrentStatusMatchingLicenseStatus
-                ? licenses.filter((license) => license.uuid !== activatedOrAutoAppliedLicense.uuid)
-                : [...licenses, activatedOrAutoAppliedLicense];
-            } else if (isCurrentStatusMatchingLicenseStatus) {
-              updatedLicensesByStatus[activatedOrAutoAppliedLicense.status].push(activatedOrAutoAppliedLicense);
-            }
-          });
-          // Optimistically update the query cache with the auto-activated or auto-applied subscription license.
-          const updatedSubscriptionLicenses = subscriptionsData.subscriptionLicenses.length > 0
-            ? subscriptionsData.subscriptionLicenses.map((license) => {
-            // Ensures an auto-activated license is updated in the query cache to change
-            // its status from "assigned" to "activated".
-              if (license.uuid === activatedOrAutoAppliedLicense.uuid) {
-                return activatedOrAutoAppliedLicense;
-              }
-              return license;
-            })
-            : [activatedOrAutoAppliedLicense];
+          const { subscriptionLicensesByStatus, subscriptionLicenses } = subscriptionsData;
+          // Create a deep copy of the structure using .map for immutability, removing
+          // the `activatedOrAutoAppliedLicense` from each list. Then, re-add the license
+          // to the correct status list.
+          const updatedLicensesByStatus = Object.fromEntries(
+            Object.entries(subscriptionLicensesByStatus).map(([key, licenses]) => [
+              key,
+              licenses.filter(
+                (existingLicense) => existingLicense.uuid !== activatedOrAutoAppliedLicense.uuid,
+              ), // Remove license immutably
+            ]),
+          );
+          if (!updatedLicensesByStatus[activatedOrAutoAppliedLicense.status]) {
+            updatedLicensesByStatus[activatedOrAutoAppliedLicense.status] = [];
+          }
+          updatedLicensesByStatus[activatedOrAutoAppliedLicense.status].push(activatedOrAutoAppliedLicense);
```

Comment:
[nit] this duplicative logic also lives in the subscriptions transform within `subscriptions.js`. Could be worth abstracting for reusability.

Secondarily, there is the `groupBy` functionality that could make this implementation more readable. Unfortunately, it is still considered experimental in the official [javascript library](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy)  but `lodash` does have an implementation as well. 

``` js
const updatedLicensesByStatus = groupBy(
  activatedOrAutoAppliedLicenses,
  (license) => license.status
);
```

Reply:
I created a util function `addLicenseToSubscriptionLicensesByStatus` for the duplicative logic. I'm going to defer on using `_.groupBy` for now, though; I believe additional refactoring would be needed to first ensure we have a valid list of `subscriptionLicenses`, included appending/updating it where appropriate, before calling `_.groupBy`.